omnipaxos/src/sequence_paxos/mod.rs

```diff
@@ -60,24 +60,22 @@ where
         let quorum = Quorum::with(config.flexible_quorum, num_nodes);
         let max_peer_pid = peers.iter().max().unwrap();
         let max_pid = *std::cmp::max(max_peer_pid, &pid) as usize;
-        let (state, leader, lds) = match &config.initial_leader {
-            Some(l) => {
-                let (role, lds) = if l.pid == pid {
-                    // we are leader in new config
-                    let mut v = vec![None; max_pid];
-                    for idx in peers.iter().map(|pid| *pid as usize - 1) {
-                        // this works as a promise
-                        v[idx] = Some(0);
-                    }
-                    (Role::Leader, Some(v))
-                } else {
-                    (Role::Follower, None)
-                };
-                let state = (role, Phase::Accept);
-                (state, *l, lds)
+        let mut outgoing = Vec::with_capacity(BUFFER_SIZE);
+        let (state, leader, lds) = match storage.get_promise().expect("Storage error") {
+            // try to do fail recovery from storage, if None then we are starting from scratch
+            Some(b) => {
+                let state = (Role::Follower, Phase::Recover);
+                for peer_pid in &peers {
+                    outgoing.push(PaxosMessage {
+                        from: pid,
+                        to: *peer_pid,
+                        msg: PaxosMsg::PrepareReq,
+                    });
+                }
+                (state, b, None)
             }
             None => {
-                match &config.skip_prepare_use_leader {
+                match &config.initial_leader {
```

Comment:
The more I see this and other parts of the code related to this, I'm leaning toward removing this feature/field in the config. I see several reasons to remove it:

1. It makes initialisation and the failure-recovery messy with promised ballot vs. initial leader.
2. It makes the sequence numbers messy with a special case first accept case.

What do you think @kevin-harrison? I would propose to remove it, since the benefit of it is quite minimal i.e., just one skipped prepare phase (which is probably very light-weight anyway since there are no entries), but the code gets much more messy.

Reply:
I definitely agree, it will make things a lot simpler. Also now that I think about it  sequence numbers mess up recovering our previous leader after storage. This is because we don't also recover the state of our sequence numbers. So as a follower even if we recover with the correct leader we will almost always falsely detect a sequence break. As a leader if we recover our ballot and are still the leader our `leader_state.follower_sequence_nums` get reset causing our followers to falsely detect sequence breaks but more importantly we will be repeating sequence numbers which is unacceptable.

Reply:
Not sure if I understand what you mean. But if a leader recovers after a crash, the idea is not for it to continue to lead in the same ballot (before it crashed). Instead, the BLE will elect a new higher ballot.

Isn't it that when a server recovers from a crash, it will go into the `Recover` mode which will reset the sequence numbers anyway (which is fine)?

Reply:
What is the purpose of recovering ballots after a crash? Is it to skip a prepare phase or for safety?